openpgp-sdkms/src/lib.rs

```diff
@@ -0,0 +1,1004 @@
+            .build()
+            .context("could not initiate an SDKMS client")?)
+    }
+}
+
+#[derive(PartialEq)]
+enum Role {
+    Signer,
+    Decryptor,
+}
+
+impl SdkmsAgent {
+    /// Returns an SdkmsAgent with signing capabilities, corresponding to the
+    /// given key name.
+    pub fn new_signer(key_name: &str) -> Result<Self> {
+        let credentials = Credentials::new_from_env()?;
+        let http_client = credentials.http_client()?;
+
+        let descriptor = SobjectDescriptor::Name(key_name.to_string());
+        let sobject = http_client
+            .get_sobject(None, &descriptor)
+            .context(format!("could not get primary key {}", key_name))?;
+        let key = PublicKey::from_sobject(sobject, KeyRole::Primary)?;
+
+        Ok(SdkmsAgent {
+            credentials,
+            descriptor,
+            public: key.sequoia_key.expect("key is not loaded"),
+            role: Role::Signer,
+        })
+    }
+
+    /// Returns an SdkmsAgent with decryption capabilities, corresponding to the
+    /// given key name.
+    pub fn new_decryptor(key_name: &str) -> Result<Self> {
+        let credentials = Credentials::new_from_env()?;
+        let http_client = credentials.http_client()?;
+
+        let descriptor = SobjectDescriptor::Name(key_name.to_string());
+        let sobject = http_client
+            .get_sobject(None, &descriptor)
+            .context(format!("could not get primary key {}", key_name))?;
+
+        if let Some(dict) = sobject.custom_metadata {
+            if dict.contains_key(SDKMS_LABEL_PGP) {
+                let json_str = &dict[SDKMS_LABEL_PGP];
+                let key_md: KeyMetadata = serde_json::from_str(json_str)?;
+                if key_md.subkeys[0].len() == 0 {
+                    return Err(Error::msg("No subkeys found in SDKMS"));
+                }
+                let uid = Uuid::parse_str(&key_md.subkeys[0])?;
+                let descriptor = SobjectDescriptor::Kid(uid);
+                let sobject = http_client
+                    .get_sobject(None, &descriptor)
+                    .context("could not get subkey".to_string())?;
+                let key = PublicKey::from_sobject(sobject, KeyRole::Subkey)?;
+                return Ok(SdkmsAgent {
+                    descriptor,
+                    public: key.sequoia_key.expect("key is not loaded"),
+                    role: Role::Decryptor,
+                    credentials,
+                });
+            }
+        }
+        Err(Error::msg("was not able to get decryption subkey"))
+    }
+}
+
+#[derive(Clone)]
+struct PublicKey {
+    role:        KeyRole,
+    descriptor:  SobjectDescriptor,
+    sequoia_key: Option<Key<PublicParts, UnspecifiedRole>>,
+}
+
+enum SupportedPkAlgo {
+    Rsa(u32),
+    Ec(ApiCurve),
+    Curve25519,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum KeyRole {
+    Primary,
+    Subkey,
+}
+
+#[derive(Serialize, Deserialize)]
+struct KeyMetadata {
+    version:     String,
```

Comment:
I don't think this is strictly required. In any case, it could be added easily in the future. Or do you think it'd be interesting to have in case (some) version of the crate/sdkms would have a serious bug that would require keys to be rotated?

Reply:
Agreed, not justified at the moment.